server/src/main/java/org/opensearch/tasks/ResourceUsageInfo.java

```diff
@@ -0,0 +1,78 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.tasks;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+import java.util.EnumMap;
+
+/**
+ * Thread resource usage information for particular resource stats type.
+ * <p>
+ * It captures the resource usage information about a particular execution of thread
+ * for a specific stats type.
+ */
+public class ResourceUsageInfo {
+    private static final Logger logger = LogManager.getLogger(ResourceUsageInfo.class);
+    private final EnumMap<ResourceStats, ResourceStatsInfo> statsInfo = new EnumMap<>(ResourceStats.class);
+
+    public ResourceUsageInfo(ResourceUsageMetric... resourceUsageMetrics) {
+        for (ResourceUsageMetric resourceUsageMetric : resourceUsageMetrics) {
+            this.statsInfo.put(resourceUsageMetric.getStats(), new ResourceStatsInfo(resourceUsageMetric.getValue()));
+        }
+    }
+
+    public void recordResourceUsageMetrics(ResourceUsageMetric... resourceUsageMetrics) {
+        for (ResourceUsageMetric resourceUsageMetric : resourceUsageMetrics) {
+            ResourceStatsInfo resourceStatsInfo = statsInfo.get(resourceUsageMetric.getStats());
+            if (resourceStatsInfo == null) {
+                statsInfo.put(resourceUsageMetric.getStats(), new ResourceStatsInfo(resourceUsageMetric.getValue()));
+            } else {
+                updateResourceUsageInfo(resourceStatsInfo, resourceUsageMetric);
+            }
+        }
+    }
```

Comment:
Seems like a concurrency bug 
1. Concurrent requests might find map to be empty and would end up instantiating fresh new resource usage
2. Its possible for `getStatsInfo ` to see a partial state which might not be coherent

Reply:
Adding a fresh entry into the map is only done during initiation. Updated in a way that ```recordResourceUsageMetrics``` only update the existing entries in the map and never adds a new entry.